### ISSUE TITLE
Question about the mIoU metric

### ISSUE THREAD

**Issue by tjyuyao** [lxasqjc/Deformation-Segmentation#4]
Dear authors,

I have carefully read your paper and have a question about the mIoU metric reported in Figure 4(a). My question is based on the following observations:

- Figure 4(a) reports `mIoU(single_loss) ~= 0.46` and `mIoU(joint_loss) ~= 0.50`, and
- I run the code in this repository with `seed=304`, and get `val_deformed_miou(single_loss) = 0.465027607866967` and `val_deformed_miou(joint_loss) = 0.49334734890825`, but  `val_miou(single_loss) = 0.418473241184434` and `val_miou(joint_loss) = 0.447862081141869`. (These values are recorded in the csv file, and the screen console seems only prints deform_miou in the name of miou)

So my question is: Does my run comply with the experiments in the paper? (Do you report deformed mIoU or mIoU of the original resolution?)

Look forward to your reply. Thanks a lot!

**Comment by lxasqjc** [comment]
Hi tiyuyao,

Thanks for your interest!

- Could you specify which code you have used to run your experiments, is it [cityscape_64_128_ours.sh](https://github.com/lxasqjc/Deformation-Segmentation/blob/main/quick_start_bash/cityscape_64_128_ours.sh)? because your results seem not consistent with what I have got.
- Regarding your question about the screen console, at the end of the training you should get both mIoU and deform_mIoU, like below, otherwise, there might be something wrong with your experiment setting, if you can provide more information I will be able to diagnose.

> Mean IoU: 0.4942, Accuracy: 78.35%, Inference Time: 92.4300s, Valid Samples: 500
Mean IoU_deformed: 0.5386, Accuracy_deformed: 76.72%, Inference Time: 92.4300s
Mean iou_y_reverse: 0.8125, Accuracy_y_reverse: 91.72%
Evaluation Done!
val_iou_deformed: 0.5386387024771562
Saving history...
Training Done!

**Comment by tjyuyao** [comment]
Hello Chen,

The following result is what I got by executing `cityscape_64_128_ours.sh` (two runs both with seed 304). 
```
Mean IoU: 0.4453, Accuracy: 78.92%, Inference Time: 47.6982s, Valid Samples: 500
Mean IoU_deformed: 0.4914, Accuracy_deformed: 80.42%, Inference Time: 47.6982s
Evaluation Done!
val_iou_deformed: 0.4914486649601866
Saving history...
```
```
Mean IoU: 0.4479, Accuracy: 78.02%, Inference Time: 96.8810s, Valid Samples: 500
Mean IoU_deformed: 0.4933, Accuracy_deformed: 78.58%, Inference Time: 96.8810s
Evaluation Done!
val_iou_deformed: 0.4933473489082504
Saving history...
Training Done!
```

I set `` to zero, to test for the single loss counterpart, and get
seed 304
```
Mean IoU: 0.4185, Accuracy: 78.06%, Inference Time: 94.4894s, Valid Samples: 500
Mean IoU_deformed: 0.4650, Accuracy_deformed: 81.20%, Inference Time: 94.4894s
Evaluation Done!
val_iou_deformed: 0.4650276078669677
```

---

I have some questions by now:

- In your paper you said you use ImageNet pretrained weights, but in the code https://github.com/lxasqjc/Deformation-Segmentation/blob/73f6f7ceb47e63b8bb1dd555520859d039a6af85/models/models.py#L2617 it says pretrained=False, so which is the case for your result `Mean IoU: 0.4942, Mean IoU_deformed: 0.5386`?
- I actually changed the content of file https://github.com/lxasqjc/Deformation-Segmentation/blob/main/data/Cityscape/training_list_2975.odgt so that I used original images (uncropped) version of CityScapes. Does this make things different? I seem not able to find an automation script for preparing the cropped dataset.
- There is another question not so much related. I noticed that the upsampling process is very slow. Does the `qhull.spatial` works faster? I find that the code is not using it.

Thanks for your time! I am currently working on a very close paper, and it would be so nice to hear from you!

Yuyao

**Comment by lxasqjc** [comment]
Hi Yuyao,

Could you pull the latest repo and rebuild environments with updated instructions as below, and let me know how it goes.
`conda env create -f deform_seg_env.yml`

> Explain: the reason is some old versions of dependencies/libs can cause decreased performance, which I had encountered during development. I have tried to rerun `cityscape_64_128_ours.sh` with our local environment which does reproduce our results. I have also tried rebuilding the environment with old `deform_seg_env.yml` and realised some incompatible dependencies occurred with updated conda/pip, hence now I updated `deform_seg_env.yml` and verified locally, preliminary results show it should reproduce similar results. 

- regarding ImageNet pre-trained weights, it was a typo in the paper and the code is correct --- no pre-trained weights are used, thanks for pointing it out.
- for [training_list_2975.odgt](https://github.com/lxasqjc/Deformation-Segmentation/blob/main/data/Cityscape/training_list_2975.odgt), they are actually the uncropped original images, please ignore the dumpted naming of '_cropped_5000_x_0_y_0', I have updated them too. 
- you are right `qhull.spatial` is not used, it was there at initial development but later dropped, hence I cannot comment on its performance, please share your findings if you got a chance to try :)

Best
Chen

**Comment by tjyuyao** [comment]
I am using the new `deform_seg_env.yml` now, but updated `torch`, `torchvision`, and `tensorboard` due to some compatibility issues of CUDA etc. after executing `conda env create -f deform_seg_env.yml`. I would update this comment once my training is finished.

**Comment by lxasqjc** [comment]
Hi Yuyao,

Thank you for your update and very glad to see you successfully reproduced the results!
My guess is it may related with how PyTorch's implementation of interpolation, something related to precision, but can not argue further :)
Anyway let me know if you got any other questions, and I will close this issue now.

Best
Chen
